### ISSUE TITLE
NULL input field generates U_ILLEGAL_ARGUMENT_ERROR

### ISSUE THREAD

**Issue by freddyatarangodb** [arangodb/arangodb#8137]
## My Environment

* __ArangoDB Version__:         3.4
* __Storage Engine__:             
* __Deployment Mode__:       
* __Deployment Strategy__:   
* __Configuration__:               
* __Infrastructure__:               
* __Operating System__:        
* __Total RAM in your machine__:        
* __Disks in use__:        
* __Used Package__:              

## Component, Query & Data

__Affected feature__:

__AQL query (if applicable)__:

__AQL explain (if applicable)__:

__Dataset__:

__Size of your Dataset on disk__:

__Replication Factor & Number of Shards (Cluster only)__:

## Steps to reproduce

1. 
2. 
3. 

__Problem__:

__Expected result__:

**Comment by freddyatarangodb** [comment]
It is a bug involving SUBSTITUTE. Basically, if the input field is NULL, it generates an U_ILLEGAL_ARGUMENT_ERROR probably from inside here https://github.com/sogko/ArangoDB/blob/master/3rdParty/icu/source/common/charstr.cpp

**Comment by dothebart** [comment]
Hi, 
You're referencing a _very_ old branch. 
our current version of ICU lives here: https://github.com/arangodb/arangodb/tree/devel/3rdParty/V8/v5.7.492.77/third_party/icu can you revalidate whether the bug still persists?
Please provide a scenario that helps us to reproduce this bug.

**Comment by fceller** [comment]
I have tried a few combinations with SUBSTITUTE in 3.4.2 without being able to reproduce.

```
arangod> db._query('return substitute("hallo", {"ll":null})').toArray()
[
  null
]
arangod> db._query('return substitute("hallo", "ll", null)').toArray()
[
  null
]
arangod> db._query('return substitute("hallo", null, null)').toArray()
[
  null
]
arangod> db._query('return substitute(null, null, null)').toArray()
[
  null
]
arangod> db._query('return substitute("hallo", {"ll":"xx"})').toArray()
[
  "haxxo"
]
```

Therefore it would be good to get the AQL that fails.

**Comment by fceller** [comment]
There is however a wrong behaivor

```
arangod> db._query('return SUBSTITUTE("the quick brown foxx", { "quick": "", "brown": null, "foxx": "ant" })').toArray()
[
  null
]
```

According to the documentation this should return "the   ant"

**Comment by jsteemann** [comment]
For me it produces "invalid argument type":
```
> db._query('return SUBSTITUTE("the quick brown foxx", { "quick": "", "brown": null, "foxx": "ant" })')
[object GeneralArrayCursor, count: 1, cached: false, warning(s): "1542 invalid argument type in call to function 'SUBSTITUTE()'"]
```
If `null` values should lead to replacing with an empty string, the fix for this should be
```
diff --git a/arangod/Aql/Functions.cpp b/arangod/Aql/Functions.cpp
index 3f0b25808b..1215a814c8 100644
--- a/arangod/Aql/Functions.cpp
+++ b/arangod/Aql/Functions.cpp
@@ -2586,12 +2586,18 @@ AqlValue Functions::Substitute(arangodb::aql::Query* query, transaction::Methods
       arangodb::velocypack::ValueLength length;
       char const* str = it.key.getString(length);
       matchPatterns.push_back(UnicodeString(str, static_cast<int32_t>(length)));
-      if (!it.value.isString()) {
+      if (it.value.isNull()) {
+        // null replacement value => replace with an empty string
+        replacePatterns.push_back(UnicodeString("", int32_t(0)));
+      } else if (it.value.isString()) {
+        // string case
+        str = it.value.getStringUnchecked(length);
+        replacePatterns.push_back(UnicodeString(str, static_cast<int32_t>(length)));
+      } else {
+        // non strings
         ::registerInvalidArgumentWarning(query, AFN);
         return AqlValue(AqlValueHintNull());
       }
-      str = it.value.getStringUnchecked(length);
-      replacePatterns.push_back(UnicodeString(str, static_cast<int32_t>(length)));
     }
   } else {
     if (parameters.size() < 2) {
```

**Comment by jsteemann** [comment]
Fixed in 3.4.3.

**Comment by sleto-it** [comment]
Hi,

Thanks for reporting this. Version 3.4.3 has been released

I am going ahead and closing this ticket now. Will you please add a new comment if further help is needed on this topic, so we can reopen the ticket?

Many Thanks,
